File: Courses/PHP Web Basics - май 2019/16.ExerciseEncapsulationInheritance/01.03.ShoppingSpree/Person.php
<?php


class Person
{
    /**
     * @var string
     */
    private $name;
    /**
     * @var float
     */
    private $money;
    /**
     * @var  Product[] $products
     */
    private $products;

    /**
     * Person constructor.
     * @param string $name
     * @param float $money
     * @throws Exception
     */
    public function __construct(string $name, float $money)
    {
        $this->setName($name);
        $this->setMoney($money);
        $this->products = [];
    }

    /**
     * @return string
     */


    public function getName(): string
    {
        return $this->name;
    }

    /**
     * @param string $name
     * @throws Exception
     */
    private function setName(string $name): void
    {
        if($name == ''){
            throw new Exception('Name cannot be empty'.PHP_EOL);
        }
        $this->name = $name;
    }

    /**
     * @return float
     */
    public function getMoney(): float
    {
        return $this->money;
    }

    /**
     * @param float $money
     * @throws Exception
     */
    private function setMoney(float $money): void
    {
        if ($money < 0) {
            throw new Exception("Money cannot be negative".PHP_EOL);
        }
        $this->money = $money;
    }

    /**
     * @return  Product[] $products
     */
    public function getProducts(): array
    {
        return $this->products;
    }

    /**
     * @param Product $product
     */
    public function addProduct(Product $product): void
    {
        $this->products[] = $product;
    }
    public function canAffortProduct(Product $product) : bool
    {
        return $product->getCost() > $this->getMoney();

    }

    /**
     * @param Product $product
     * @throws Exception
     */
    public function buyProduct(Product $product) : void
    {
        if ($this->canAffortProduct($product)) {

            throw new Exception("{$this->getName()} can't afford {$product->getName()}".PHP_EOL);
        }
        $this->setMoney($this->getMoney() - $product->getCost());
        $this->addProduct($product);
        echo "{$this->getName()} bought {$product->getName()}".PHP_EOL;
    }
    public function __toString()
    {
        if (count($this->getProducts()) === 0) {
           return $this->getName(). " - Nothing bought".PHP_EOL;
        }

        $output =  $this->getName(). " - ";

        $output .= implode(",", array_map(function (Product $product) {
            return $product->getName();
        }, $this->getProducts()));
        return $output.PHP_EOL;
    }
}

File: Courses/C# OOP/Basics/08. Encapsulation Exercise/08. Encapsulation Exercise/04.ShoppingSpree/Person.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace ShoppingSpree
{
    public class Person
    {
        private string name;
        private decimal money;
        private List<Product> products;

        public Person(string name, decimal money)
        {
            Name = name;
            Money = money;
            Products = new List<Product>();
        }

        public string Name 
        { 
            get => name;

            set
            {
                if (string.IsNullOrEmpty(value) || string.IsNullOrWhiteSpace(value))
                {
                    throw new ArgumentException("Name cannot be empty");
                }
                name = value;
            }
        }
        public decimal Money
        { 
            get => money;
            set
            {
                if (value < 0)
                {
                    throw new ArgumentException("Money cannot be negative");
                }
                money = value;
            }
        }
        public List<Product> Products { get => products; set => products = value; }

        public void AddProduct(Product product)
        {
            decimal cost = product.Cost;

            if (cost > this.Money)
            {
                Console.WriteLine($"{Name} can't afford {product.Name}");
            }
            else
            {
                Products.Add(product);
                Money -= cost;
                Console.WriteLine($"{Name} bought {product.Name}");
            }
        }

        public override string ToString()
        {
            if (Products.Count == 0)
            {
                return $"{Name} - Nothing bought";
            }
            return $"{Name} - {string.Join(", ", Products.Select(x => x.ToString()))}";
        }
    }
}
